examples/pallas/blocksparse_matmul.py

```diff
@@ -0,0 +1,202 @@
+    self.shape = shape
+
+  def tree_flatten(self):
+    return (self.blocks, self.blocks_per_row, self.indices), (self.shape,)
+
+  @classmethod
+  def tree_unflatten(cls, data, xs):
+    blocks, blocks_per_row, indices = xs
+    shape, = data
+    return BlockELL(blocks, blocks_per_row, indices, shape=shape)
+
+  def _validate(self):
+    nblocks, n, m = self.blocks.shape
+    nrows = self.blocks_per_row.shape[0]
+    assert self.indices.shape[0] == nrows
+    assert len(self.shape) == 2
+    assert self.shape[0] == n * nrows
+    assert self.shape[1] % m == 0
+  
+  @jax.jit
+  def todense(self) -> jnp.ndarray:
+    self._validate()
+    _, n, m = self.blocks.shape
+    nrows = self.shape[0] // n
+    out = jnp.zeros(self.shape, self.dtype)
+    def i_body(i, out):
+      num_blocks_in_row = self.blocks_per_row[i]
+      def j_body(j, out):
+        jj, sparse_idx = self.indices[i, j]
+        out = lax.dynamic_update_slice(out, self.blocks[sparse_idx], (i * n, jj * m))
+        return out
+      return lax.fori_loop(0, num_blocks_in_row, j_body, out)
+    return lax.fori_loop(0, nrows, i_body, out)
+
+  @jax.jit
+  def __matmul__(self, other):
+    assert isinstance(other, jnp.ndarray)
+    self._validate()
+    assert self.ndim == other.ndim == 2
+    assert self.shape[1] == other.shape[0]
+    out = jnp.zeros((self.shape[0], other.shape[1]),
+                    dtype=jnp.result_type(self.dtype, other.dtype))
+    _, n, m = self.blocks.shape
+    nrows = self.shape[0] // n
+    def i_body(i, out):
+      num_blocks_in_row = self.blocks_per_row[i]
+      acc = jnp.zeros((n, other.shape[1]), dtype=jnp.float32)
+      def k_body(k, acc):
+        jj, sparse_idx = self.indices[i, k]
+        chunk = lax.dynamic_slice(other, [jj * m, 0], (m, other.shape[1]))
+        block = self.blocks[sparse_idx]
+        return acc + block.dot(chunk)
+      acc = lax.fori_loop(0, num_blocks_in_row, k_body, acc).astype(out.dtype)
+      return lax.dynamic_update_slice(out, acc, (i, 0))
+    return lax.fori_loop(0, nrows, i_body, out)
+
+def sample_sparse_matrix(key, m, n, bm, bn, *, sparse_prob=0.2,
+                         dtype=jnp.float32) -> BlockELL:
+  k1, k2, k3 = random.split(key, 3)
+  num_rows = m // bm
+  num_cols = n // bn
+  blocks_per_row = tfd.Binomial(num_cols, probs=sparse_prob).sample(
+      seed=k1, sample_shape=[num_rows]).astype(jnp.int32)
+  num_blocks = blocks_per_row.sum()
+  indices = []
+  block_index = 0
+  max_num_blocks = blocks_per_row.max(axis=0)
+  for i, k in zip(range(num_rows), random.split(k2, num_rows)):
+    row = []
+    num_blocks_in_row = blocks_per_row[i]
+    block_indices = jnp.sort(random.permutation(k, jnp.arange(num_cols))[:max_num_blocks])
+    for j, b in zip(range(max_num_blocks), block_indices):
+      if j < num_blocks_in_row:
+        index = [b, block_index]
+        block_index += 1
+      else:
+        index = [0, 0]
+      row.append(index)
+    indices.append(row)
+  indices = jnp.array(indices)
```

Comment:
Am I reading this right that `indices` is three dimensions? How does this square with the index shape being `[n_rows, n_blocks]` above?

Reply:
Oops that comment is wrong. `indices` is `int32[n_rows, max_num_blocks_per_row, 2]`